### PR TITLE
Cache temp files created from base64 data

### PR DESCRIPTION
**Pull request by abidlabs**
This PR slightly changes the way that we generate temp files from base64 input data by giving them a deterministic path based on the contents of data. In other words, if a user submits the same input twice, it will be produce the same temp file. If the temp file already exists, then the conversion from base64 to binary (which is very time consuming) can be skipped and the file can be returned directly. 

Fixes: #3189

Probably the easiest way to test this would be to run a demo like this:

```py
import gradio as gr

gr.Interface(lambda x:x, gr.Video(source="webcam"), gr.Video()).launch()
```

* Record a video that is at least a few seconds long
* Click the submit button (note the inference time)
* Click the submit button again with the same video (note the inference time, should be much faster on this branch)

### CODE REVIEW

File: test/test_processing_utils.py [hunk 1]
```diff
@@ -194,6 +194,30 @@ def test_make_temp_copy_if_needed(self, mock_copy):
         )
         assert len(temp_file_manager.temp_files) == 2
 
+    def test_base64_to_temp_file_if_needed(self):
```
Comment (freddyaboulton) [inline test/test_processing_utils.py:197]: I would test the case of the same audio file but with different min/max values. That's an interesting case I would not have thought of!

**Review by freddyaboulton — APPROVED** [review]
Thanks for the fix @abidlabs !

File: gradio/components.py [hunk 2]
```diff
@@ -2041,20 +2037,26 @@ def preprocess(
...
 
         sample_rate, data = processing_utils.audio_from_file(
             temp_file_path, crop_min=crop_min, crop_max=crop_max
         )
 
+        # Need a unique name for the file to avoid re-using the same audio file if
+        # a user submits the same audio file twice, but with different crop min/max.
+        temp_file_path = Path(temp_file_path)
+        output_file_name = str(
+            temp_file_path.with_name(
```
[the remark this answers is not in the document]
Comment (abidlabs) [reply gradio/components.py:2050]: I had originally used `with_stem()` but turns out it's not compatible with Python 3.7-3.8: https://stackoverflow.com/a/67612849/5209347

File: test/test_processing_utils.py (hunk 1 above)
Comment (abidlabs) [reply test/test_processing_utils.py:197]: Good idea thanks!

**Comment by aliabid94** [comment]
As I understand it, on the frontend side, we still have to upload any file. The only step we save is on the backend, where we don't have to convert the base64 stream back to a binary file.

While this PR does improve things, a better approach would be to do caching on the frontend side. Because right now, let's say you have a live interface with a Video and Slider input. If you attach a video, as you move the slider, the video gets sent every single time you move the slider. If the frontend can keep track of previously uploaded files and the file names of those in the backend (which should be very feasible with PR #3191), then we only even send the file to the backend if its a new file, which saves a lot more processing and network time.

**Comment by abidlabs** [comment]
Agreed @aliabid94 that would be the optimal solution. However, which components are we planning on doing this large file upload caching with? I thought we were only applying it to the `File` and `UploadButton` components -- is it reasonable to apply the same approach to all of the other components affected by this PR: `Video`, `Audio`, `Model3D`?

File: test/test_processing_utils.py (hunk 1 above)
Comment (abidlabs) [reply test/test_processing_utils.py:197]: Added a test in `test/test_components`

**Comment by abidlabs** [comment]
Synced with @aliabid94 and we'll merge this in while @aliabid94 works on a proper fix via the upload logic in #3191

**Comment by abidlabs** [comment]
Thanks for the review and feedback guys!

**Pull request merged**